### ISSUE TITLE
Allow pre-sizing for appropriate ImmutableCollection builders

### ISSUE THREAD

**Issue by gissuebot** [google/guava#196]
_[Original issue](https://code.google.com/p/guava-libraries/issues/detail?id=196) created by **lindner+unus...@inuus.com** on 2009-06-23 at 01:15 AM_

---

It might be nice to initialize the temporary arraylist used for Immutable 
classes to a specific size by adding an optional constructor that takes sizing 
parameters.

**Comment by cowwoc** [comment]
A simple workaround is to use a mutable list to store intermediate results. Example:

```
List<String> temp = new ArrayList<>(expectedSize);

// populate "temp" ...

ImmutableList result = ImmutableList.copyOf(temp);
```

I assume that this will result in better performance than using `ImmutableList.builder()` for the same operation.

**Comment by BrandonArp** [comment]
+1  We've been waiting for this for a while.  Any movement?

**Comment by magnet** [comment]
+1, any status on this?

**Comment by NMarrero** [comment]
+1 Excited to hear about an update!

**Comment by cozmic72** [comment]
+1 Please put this in the next release.

**Comment by kevinb9n** [comment]
We have actually been sinking a fair amount of time into studying this.

At every turn we keep finding that the actual measurable benefits are significant in *fewer circumstances* than we thought, and that those benefits are *smaller* than we thought.

We're not denying that those benefits exist. But the costs are also real; there IS an added burden that many thousands of users will take on once a tuning parameter like this were to exist - deciding whether to use it, discussing with a reviewer whether to use it, adding bulk to their code to use it, actually putting thought into how to come up with a "good guess" when the real value isn't known (note that is not the common case), etc. That may not seem like a big cost, but it's not free.

Our job before adding new APIs to Guava is to make damned sure the benefits will outweigh the costs, and we may actually get there. I guess I'm just explaining why this is not a no-brainer and why it is taking us some time.

Also, I'm sorry that it is now too late for 19.0, as rc1 is already out.

**Comment by kevinb9n** [comment]
I do realize that on 2012-12-20 I announced that we had decided in favor of this. I'm sorry that decision eroded. That's pretty annoying.

**Comment by schlosna** [comment]
Looks like some things landed in master, so maybe in Guava 24.0 release TBD?

- [`ImmutableList.builderWithExpectedSize`](https://github.com/google/guava/commit/7928bbe079303c54373453353c9ef2cd7de9365e): https://github.com/google/guava/commit/7928bbe079303c54373453353c9ef2cd7de9365e
- [`Immutable{Map,BiMap}.builderWithExpectedSize`](https://github.com/google/guava/commit/8158909d7bda9f98a4576f89eb5fd88cb12c2bd6): https://github.com/google/guava/commit/8158909d7bda9f98a4576f89eb5fd88cb12c2bd6
- [`ImmutableSet.builderForExpectedSize`](https://github.com/google/guava/commit/d7501abccbd666a8db4dff77fa110c801a094c75): https://github.com/google/guava/commit/d7501abccbd666a8db4dff77fa110c801a094c75

**Comment by kluever** [comment]
Phew, I had a minor heart attack when I saw those commit titles, but it turns out we did indeed use `builderWithExpectedSize()` consistently for all of the APIs, and did _not_ use `builder**For**ExpectedSize()` for `ImmutableSet`.
